src/App.spec.ts

```diff
@@ -342,18 +347,129 @@ describe('App', () => {
         assert.instanceOf(fakeErrorHandler.firstCall.args[0], Error);
       });
+
+      it('correctly waits for async listeners', async () => {
+        let changed = false;
+
+        app.use(async ({ next }) => {
+          await delay(100);
+          changed = true;
+
+          await next();
+        });
+
+        await fakeReceiver.sendEvent(dummyReceiverEvent);
+        assert.isTrue(changed);
+        assert(fakeErrorHandler.notCalled);
+      });
+
+      it('throws errors which can be caught by upstream async listeners', async () => {
+        const thrownError = new Error('Error handling the message :(');
+        let caughtError;
+
+        app.use(async ({ next }) => {
+          try {
+            await next();
+          } catch (err) {
+            caughtError = err;
+          }
+        });
+
+        app.use(async () => {
+          throw thrownError;
+        });
+
+        app.error(fakeErrorHandler);
+
+        await fakeReceiver.sendEvent(dummyReceiverEvent);
+
+        assert.equal(caughtError, thrownError);
+        assert(fakeErrorHandler.notCalled);
+      });
+
+      it('calls async middleware in declared order', async () => {
+        const message = ':wave:';
+        let middlewareCount = 0;
+
+        /**
+         * Middleware that, when calls, asserts that it was called in the correct order
+         * @param orderDown The order it should be called when processing middleware down the stack
+         * @param orderUp The order it should be called when processing middleware up the stack
+         */
+        const assertOrderMiddleware = (orderDown: number, orderUp: number) => async ({ next }: any) => {
+          await delay(100);
+          middlewareCount += 1;
+          assert.equal(middlewareCount, orderDown);
+          await next();
+          middlewareCount += 1;
+          assert.equal(middlewareCount, orderUp);
+        };
+
+        app.use(assertOrderMiddleware(1, 8));
+        app.message(message, assertOrderMiddleware(3, 6), assertOrderMiddleware(4, 5));
+        app.use(assertOrderMiddleware(2, 7));
+        app.error(fakeErrorHandler);
+
+        await fakeReceiver.sendEvent({
+          ...dummyReceiverEvent,
+          body: {
+            type: 'event_callback',
+            event: {
+              type: 'message',
+              text: message,
+            },
+          },
+        });
+
+        assert.equal(middlewareCount, 8);
+        assert(fakeErrorHandler.notCalled);
+      });
+
+      it('should, on error call the global error handler', async () => {
+        const error = new Error('Everything is broke, you probably should restart, if not then good luck');
+
+        app.use(() => {
+          throw error;
+        });
+
+        app.error(async (actualError) => {
+          assert.instanceOf(actualError, UnknownError);
+          assert.equal(actualError.message, error.message);
+          await delay(); // Make this async to make sure error handlers can be tested
```

Comment:
what happens if we remove this line? since the function is declared an `async` function, i think the behavior we intend to test remains either way.